From: Richard Hansen <rhansen@rhansen.org>
Date: Sun, 22 Mar 2026 17:07:41 -0400
Subject: estimate: Redo searches for different major version

Use `golang.org/x/mod/module.SplitPathVersion` to find related Go
modules that have already been packged in Debian.  This is a bit more
robust than parsing the module paths ourselves, as it handles corner
cases such as gopkg.in version suffixes.

Also, return the "closest" match, defined as the Debian package that
provides the module whose major version is most similar to the desired
major version (with ties broken in favor of the older version).
---
 estimate.go | 95 ++++++++++++++++++++++++++++++++++++++++++-------------------
 1 file changed, 66 insertions(+), 29 deletions(-)

diff --git a/estimate.go b/estimate.go
index 15dfe44..1c833ad 100644
--- a/estimate.go
+++ b/estimate.go
@@ -16,15 +16,13 @@ import (
 	"strings"
 
 	"github.com/mattn/go-isatty"
+	"golang.org/x/mod/module"
 )
 
 const (
 	sourcesInNewURL = "https://api.ftp-master.debian.org/sources_in_suite/new"
 )
 
-// majorVersionRegexp checks if an import path contains a major version suffix.
-var majorVersionRegexp = regexp.MustCompile(`([/.])v([0-9]+)$`)
-
 // moduleBlocklist is a map of modules that we want to exclude from the estimate
 // output, associated with the reason why.
 var moduleBlocklist = map[string]string{
@@ -142,35 +140,61 @@ func getDirectDependencies(gopath, repodir, module string) (map[string]bool, err
 	return deps, nil
 }
 
-// otherVersions guesses the import paths of potential other major version
-// of the given module import path, based on [majorVersionRegex].
-func otherVersions(mod string) (mods []string) {
-	matches := majorVersionRegexp.FindStringSubmatch(mod)
-	if matches == nil {
-		return
-	}
-	matchFull, matchSep, matchVer := matches[0], matches[1], matches[2]
-	matchIndex := len(mod) - len(matchFull)
-	prefix := mod[:matchIndex]
-	version, _ := strconv.Atoi(matchVer)
-	for v := version - 1; v > 1; v-- {
-		mods = append(mods, prefix+matchSep+"v"+strconv.Itoa(v))
-	}
-	mods = append(mods, prefix)
-	return
+// splitMajorVersion is like [module.SplitPathVersion] except it returns the major version suffix as
+// an integer instead of a string.  1 is returned for all suffix-less module paths.
+func splitMajorVersion(modPath string) (string, int, error) {
+	if err := module.CheckPath(modPath); err != nil {
+		return "", 0, err
+	}
+	pfx, majSfx, _ := module.SplitPathVersion(modPath)
+	if majSfx == "" {
+		return modPath, 1, nil
+	}
+	majStr, ok := strings.CutPrefix(majSfx, "/v")
+	if !ok {
+		majStr, ok = strings.CutPrefix(majSfx, ".v")
+		if !ok {
+			panic("major version suffix does not match \"/v*\" or \".v*\"")
+		}
+	}
+	maj, err := strconv.Atoi(majStr)
+	if err != nil {
+		return "", 0, err
+	}
+	return pfx, maj, nil
 }
 
-// findOtherVersion search in m for potential other versions of the given
-// module and returns the number of the major version found, 0 if not,
-// along with the corresponding package name.
-func findOtherVersion(m map[string]debianPackage, mod string) (int, debianPackage) {
-	versions := otherVersions(mod)
-	for i, version := range versions {
-		if pkg, ok := m[version]; ok {
-			return len(versions) - i, pkg
+// bestOtherMaj searches for a Debian package that provides mod except at a different major version.
+// If one is found, the other major version and Debian package are returned.  If no match is found,
+// zero values are returned.  1 is returned for module paths without a major version suffix even if
+// the major version is v0.
+func bestOtherMaj(m map[string]map[string]debianPackage, mod string) (int, debianPackage) {
+	pfx, maj, err := splitMajorVersion(mod)
+	if err != nil {
+		log.Printf("WARNING: failed to extract major version from module path %v: %v", mod, err)
+		return 0, debianPackage{}
+	}
+	best := 0
+	bestPkg := debianPackage{}
+	bestDist := 0
+	for other, otherPkg := range m[pfx] {
+		if other == mod {
+			continue
+		}
+		_, otherMaj, err := splitMajorVersion(other)
+		if err != nil {
+			log.Printf("WARNING: failed to extract major version from module path %v: %v", other, err)
+			continue
+		}
+		diff := otherMaj - maj
+		otherDist := max(diff, -diff)
+		if best == 0 || otherDist < bestDist || (otherDist == bestDist && otherMaj < maj) {
+			best = otherMaj
+			bestPkg = otherPkg
+			bestDist = otherDist
 		}
 	}
-	return 0, debianPackage{}
+	return best, bestPkg
 }
 
 // trackerLink generates an OSC 8 hyperlink to the tracker for the given Debian
@@ -239,6 +263,19 @@ func estimate(importpath, revision string) error {
 	if err != nil {
 		return fmt.Errorf("get golang debian packages: %w", err)
 	}
+	packagedByModPathPrefix := map[string]map[string]debianPackage{}
+	for modPath, pkg := range golangBinaries {
+		pfx, _, ok := module.SplitPathVersion(modPath)
+		if !ok {
+			log.Printf("WARNING: ignoring invalid module path %v for Debian package %v",
+				modPath, pkg.binary)
+			continue
+		}
+		if packagedByModPathPrefix[pfx] == nil {
+			packagedByModPathPrefix[pfx] = map[string]debianPackage{}
+		}
+		packagedByModPathPrefix[pfx][modPath] = pkg
+	}
 	sourcesInNew, err := getSourcesInNew()
 	if err != nil {
 		return fmt.Errorf("get packages in new: %w", err)
@@ -324,7 +361,7 @@ func estimate(importpath, revision string) error {
 				return // already packaged in Debian
 			}
 			// Check for potential other major versions already in Debian.
-			v, pkg := findOtherVersion(golangBinaries, mod)
+			v, pkg := bestOtherMaj(packagedByModPathPrefix, mod)
 			if v != 0 {
 				// Log info to indicate that it is an approximate match
 				// but consider that it is packaged and skip the children.
